File: jvb/src/test/java/org/jitsi/videobridge/rest/root/colibri/mucclient/MucClientTest.java
/*
 * Copyright @ 2018 - present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jitsi.videobridge.rest.root.colibri.mucclient;

import org.eclipse.jetty.http.*;
import org.glassfish.jersey.server.*;
import org.glassfish.jersey.test.*;
import org.jitsi.videobridge.rest.*;
import org.jitsi.videobridge.util.*;
import org.jitsi.videobridge.xmpp.*;
import org.json.simple.*;
import org.junit.*;
import org.mockito.*;

import javax.ws.rs.client.*;
import javax.ws.rs.core.*;
import javax.ws.rs.core.Application;

import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

public class MucClientTest extends JerseyTest
{
    protected ClientConnectionImplSupplier clientConnectionImplSupplier;
    protected ClientConnectionImpl clientConnection;
    protected static final String BASE_URL = "/colibri/muc-client";

    @Override
    protected Application configure()
    {
        clientConnectionImplSupplier = mock(ClientConnectionImplSupplier.class);
        clientConnection = mock(ClientConnectionImpl.class);
        when(clientConnectionImplSupplier.get()).thenReturn(clientConnection);

        enable(TestProperties.LOG_TRAFFIC);
        enable(TestProperties.DUMP_ENTITY);
        return new ResourceConfig() {
            {
                register(new MockBinder<>(clientConnectionImplSupplier, ClientConnectionImplSupplier.class));
                register(MucClient.class);
            }
        };
    }

    @SuppressWarnings("unchecked")
    @Test
    public void testAddMuc()
    {
        ArgumentCaptor<JSONObject> jsonConfigCaptor =
                ArgumentCaptor.forClass(JSONObject.class);

        when(clientConnection.addMucClient(jsonConfigCaptor.capture())).thenReturn(true);
        JSONObject json = new JSONObject();
        json.put("id", "id");
        json.put("hostname", "hostname");
        json.put("username", "username");
        json.put("password", "password");
        json.put("muc_jids", "jid1, jid2");
        json.put("muc_nickname", "muc_nickname");

        Response resp = target(BASE_URL + "/add").request().post(Entity.json(json.toJSONString()));
        assertEquals(HttpStatus.OK_200, resp.getStatus());
        assertEquals(json, jsonConfigCaptor.getValue());
    }

    @SuppressWarnings("unchecked")
    @Test
    public void testAddMucFailure()
    {
        when(clientConnection.addMucClient(any())).thenReturn(false);
        JSONObject json = new JSONObject();
        json.put("id", "id");
        json.put("hostname", "hostname");
        json.put("username", "username");
        json.put("password", "password");
        json.put("muc_jids", "jid1, jid2");
        json.put("muc_nickname", "muc_nickname");

        Response resp = target(BASE_URL + "/add").request().post(Entity.json(json.toJSONString()));
        assertEquals(HttpStatus.BAD_REQUEST_400, resp.getStatus());
    }

    @SuppressWarnings("unchecked")
    @Test
    public void testRemoveMuc()
    {
        ArgumentCaptor<JSONObject> jsonConfigCaptor =
                ArgumentCaptor.forClass(JSONObject.class);

        when(clientConnection.removeMucClient(jsonConfigCaptor.capture())).thenReturn(true);
        JSONObject json = new JSONObject();
        json.put("id", "id");

        Response resp = target(BASE_URL + "/remove").request().post(Entity.json(json.toJSONString()));
        assertEquals(HttpStatus.OK_200, resp.getStatus());
        assertEquals(json, jsonConfigCaptor.getValue());
    }

    @SuppressWarnings("unchecked")
    @Test
    public void testRemoveMucFailure()
    {
        when(clientConnection.removeMucClient(any())).thenReturn(false);
        JSONObject json = new JSONObject();
        json.put("id", "id");

        Response resp = target(BASE_URL + "/remove").request().post(Entity.json(json.toJSONString()));
        assertEquals(HttpStatus.BAD_REQUEST_400, resp.getStatus());
    }
}

File: jvb/src/test/kotlin/org/jitsi/videobridge/rest/root/colibri/mucclient/MucClientTest.kt
/*
 * Copyright @ 2018 - present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jitsi.videobridge.rest.root.colibri.mucclient

import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import org.eclipse.jetty.http.HttpStatus
import org.glassfish.jersey.server.ResourceConfig
import org.glassfish.jersey.test.JerseyTest
import org.glassfish.jersey.test.TestProperties
import org.jitsi.videobridge.rest.MockBinder
import org.jitsi.videobridge.xmpp.XmppConnection
import org.json.simple.JSONObject
import org.junit.Test
import javax.ws.rs.client.Entity
import javax.ws.rs.core.Application

class MucClientTest : JerseyTest() {
    private lateinit var xmppConnection: XmppConnection
    private val baseUrl = "/colibri/muc-client"

    override fun configure(): Application {
        xmppConnection = mockk()

        enable(TestProperties.LOG_TRAFFIC)
        enable(TestProperties.DUMP_ENTITY)
        return object : ResourceConfig() {
            init {
                register(MockBinder(xmppConnection, XmppConnection::class.java))
                register(MucClient::class.java)
            }
        }
    }

    @Test
    fun testAddMuc() {
        val jsonConfigSlot = slot<JSONObject>()

        every { xmppConnection.addMucClient(capture(jsonConfigSlot)) } returns true

        val json = JSONObject().apply {
            put("id", "id")
            put("hostname", "hostname")
            put("username", "username")
            put("password", "password")
            put("muc_jids", "jid1, jid2")
            put("muc_nickname", "muc_nickname")
        }

        val resp = target("$baseUrl/add").request().post(Entity.json(json.toJSONString()))
        resp.status shouldBe HttpStatus.OK_200
        jsonConfigSlot.captured shouldBe json
    }

    @Test
    fun testAddMucFailure() {
        val jsonConfigSlot = slot<JSONObject>()

        every { xmppConnection.addMucClient(capture(jsonConfigSlot)) } returns false

        val json = JSONObject().apply {
            put("id", "id")
            put("hostname", "hostname")
            put("username", "username")
            put("password", "password")
            put("muc_jids", "jid1, jid2")
            put("muc_nickname", "muc_nickname")
        }

        val resp = target("$baseUrl/add").request().post(Entity.json(json.toJSONString()))
        resp.status shouldBe HttpStatus.BAD_REQUEST_400
    }

    @Test
    fun testRemoveMuc() {
        val jsonConfigSlot = slot<JSONObject>()

        every { xmppConnection.addMucClient(capture(jsonConfigSlot)) } returns true

        val json = JSONObject().apply {
            put("id", "id")
        }

        val resp = target("$baseUrl/add").request().post(Entity.json(json.toJSONString()))
        resp.status shouldBe HttpStatus.OK_200
        jsonConfigSlot.captured shouldBe json
    }

    @Test
    fun testRemoveMucFailure() {
        every { xmppConnection.addMucClient(any()) } returns false

        val json = JSONObject().apply {
            put("id", "id")
        }
        val resp = target("$baseUrl/add").request().post(Entity.json(json.toJSONString()))
        resp.status shouldBe HttpStatus.BAD_REQUEST_400
    }
}
